File: mock-src/typescript/src/services/TaskService.ts
/**
 * Task service for managing tasks.
 */

import { Task, TaskStatus, TaskPriority, ITask } from '../models/Task';
import { validateTaskTitle } from '../utils/validators';

/**
 * Service error base class.
 */
export class ServiceError extends Error {
  constructor(
    message: string,
    public code: string = 'SERVICE_ERROR'
  ) {
    super(message);
    this.name = 'ServiceError';
  }
}

/**
 * Not found error.
 */
export class NotFoundError extends ServiceError {
  constructor(entityType: string, entityId: string) {
    super(`${entityType} with ID ${entityId} not found`, 'NOT_FOUND');
    this.name = 'NotFoundError';
  }
}

/**
 * Validation error.
 */
export class ValidationError extends ServiceError {
  constructor(
    public field: string,
    message: string
  ) {
    super(`Validation failed for ${field}: ${message}`, 'VALIDATION_ERROR');
    this.name = 'ValidationError';
  }
}

/**
 * Options for creating a task.
 */
export interface CreateTaskOptions {
  description?: string;
  assigneeId?: string;
  priority?: TaskPriority;
  dueDate?: Date;
  tags?: string[];
}

/**
 * Options for updating a task.
 */
export interface UpdateTaskOptions {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  assigneeId?: string;
  dueDate?: Date;
}

/**
 * Service for task management operations.
 *
 * Handles business logic for creating, updating, and querying tasks.
 */
export class TaskService {
  private tasks: Map<string, Task> = new Map();
  private cache: Map<string, Task> = new Map();

  /**
   * Create a new task.
   *
   * @param title - Task title
   * @param projectId - ID of the project the task belongs to
   * @param options - Additional task options
   * @returns The created task
   * @throws ValidationError if title is invalid
   */
  createTask(
    title: string,
    projectId: string,
    options: CreateTaskOptions = {}
  ): Task {
    const validation = validateTaskTitle(title);
    if (!validation.isValid) {
      throw new ValidationError('title', validation.errorMessage!);
    }

    const task = new Task(title, projectId, {
      description: options.description,
      assigneeId: options.assigneeId,
      priority: options.priority,
      dueDate: options.dueDate,
      tags: options.tags,
    });

    this.tasks.set(task.id, task);
    return task;
  }

  /**
   * Get a task by ID.
   *
   * @param taskId - The task's unique identifier
   * @returns The task
   * @throws NotFoundError if task doesn't exist
   */
  getTask(taskId: string): Task {
    const task = this.tasks.get(taskId);
    if (!task) {
      throw new NotFoundError('Task', taskId);
    }
    return task;
  }

  /**
   * Update a task's properties.
   *
   * @param taskId - The task to update
   * @param options - Properties to update
   * @returns The updated task
   * @throws NotFoundError if task doesn't exist
   * @throws ValidationError if new values are invalid
   */
  updateTask(taskId: string, options: UpdateTaskOptions): Task {
    const task = this.getTask(taskId);

    if (options.title !== undefined) {
      const validation = validateTaskTitle(options.title);
      if (!validation.isValid) {
        throw new ValidationError('title', validation.errorMessage!);
      }
      task.title = options.title;
    }

    if (options.description !== undefined) {
      task.description = options.description;
    }

    if (options.status !== undefined) {
      task.status = options.status;
    }

    if (options.priority !== undefined) {
      task.priority = options.priority;
    }

    if (options.assigneeId !== undefined) {
      task.assigneeId = options.assigneeId;
    }

    if (options.dueDate !== undefined) {
      task.dueDate = options.dueDate;
    }

    task.updatedAt = new Date();
    this.invalidateCache(taskId);
    return task;
  }

  /**
   * Delete a task.
   *
   * @param taskId - The task to delete
   * @returns True if deleted
   * @throws NotFoundError if task doesn't exist
   */
  deleteTask(taskId: string): boolean {
    if (!this.tasks.has(taskId)) {
      throw new NotFoundError('Task', taskId);
    }

    this.tasks.delete(taskId);
    this.invalidateCache(taskId);
    return true;
  }

  /**
   * Create a task asynchronously.
   *
   * @param title - Task title
   * @param projectId - Project ID
   * @param options - Additional task options
   * @returns Promise resolving to the created task
   */
  async createTaskAsync(
    title: string,
    projectId: string,
    options: CreateTaskOptions = {}
  ): Promise<Task> {
    await new Promise((resolve) => setTimeout(resolve, 10));
    return this.createTask(title, projectId, options);
  }

  /**
   * Create multiple tasks asynchronously.
   *
   * @param tasksData - List of task data
   * @returns Promise resolving to list of created tasks
   */
  async bulkCreateAsync(
    tasksData: Array<{ title: string; projectId: string } & CreateTaskOptions>
  ): Promise<Task[]> {
    const tasks: Task[] = [];
    for (const data of tasksData) {
      const task = await this.createTaskAsync(
        data.title,
        data.projectId,
        data
      );
      tasks.push(task);
    }
    return tasks;
  }

  /**
   * Assign a task to a user.
   *
   * @param taskId - The task to assign
   * @param userId - The user to assign to
   * @returns The updated task
   */
  assignTask(taskId: string, userId: string): Task {
    const task = this.getTask(taskId);
    task.assignTo(userId);
    return task;
  }

  /**
   * Mark a task as completed.
   *
   * @param taskId - The task to complete
   * @returns The updated task
   */
  completeTask(taskId: string): Task {
    const task = this.getTask(taskId);
    task.markComplete();
    return task;
  }

  /**
   * Get all tasks for a specific project.
   *
   * @param projectId - The project ID to filter by
   * @returns List of tasks belonging to the project
   */
  getTasksByProject(projectId: string): Task[] {
    return Array.from(this.tasks.values()).filter(
      (t) => t.projectId === projectId
    );
  }

  /**
   * Get all tasks assigned to a specific user.
   *
   * @param userId - The user ID to filter by
   * @returns List of tasks assigned to the user
   */
  getTasksByAssignee(userId: string): Task[] {
    return Array.from(this.tasks.values()).filter(
      (t) => t.assigneeId === userId
    );
  }

  /**
   * Get all tasks with a specific status.
   *
   * @param status - The status to filter by
   * @returns List of tasks with the given status
   */
  getTasksByStatus(status: TaskStatus): Task[] {
    return Array.from(this.tasks.values()).filter((t) => t.status === status);
  }

  /**
   * Get all overdue tasks.
   *
   * @returns List of tasks past their due date
   */
  getOverdueTasks(): Task[] {
    return Array.from(this.tasks.values()).filter((t) => t.isOverdue);
  }

  /**
   * Search tasks with filters.
   *
   * @param query - Search text
   * @param projectId - Optional project filter
   * @param status - Optional status filter
   * @returns Matching tasks
   */
  searchTasks(
    query: string,
    projectId?: string,
    status?: TaskStatus
  ): Task[] {
    const queryLower = query.toLowerCase();

    return Array.from(this.tasks.values()).filter((task) => {
      const matchesQuery =
        task.title.toLowerCase().includes(queryLower) ||
        task.description.toLowerCase().includes(queryLower);

      if (!matchesQuery) return false;
      if (projectId && task.projectId !== projectId) return false;
      if (status && task.status !== status) return false;

      return true;
    });
  }

  /**
   * Get total number of tasks.
   */
  get taskCount(): number {
    return this.tasks.size;
  }

  /**
   * Calculate a priority score for sorting.
   *
   * @param task - The task to score
   * @returns Priority score (higher = more important)
   */
  static calculatePriorityScore(task: Task): number {
    let score = task.priority * 10;

    if (task.isOverdue) {
      score += 50;
    }

    if (task.status === TaskStatus.BLOCKED) {
      score -= 20;
    }

    return score;
  }

  private invalidateCache(taskId: string): void {
    this.cache.delete(taskId);
  }
}

File: mock-src/python/tasktracker/services/task_service.py
"""Task service for managing tasks."""

import asyncio
from datetime import datetime
from typing import Optional
from uuid import UUID

from ..models.task import Task, TaskStatus, TaskPriority
from ..repositories.task_repository import TaskRepository
from ..utils.validators import validate_task_title
from .base import (
    BaseService,
    NotFoundError,
    ValidationError,
    log_call,
    measure_time,
)


class TaskService(BaseService):
    """
    Service for task management operations.

    Handles business logic for creating, updating, and querying tasks.
    Uses TaskRepository for data persistence.

    Attributes:
        repository: The task repository for data access.
        _cache: Simple cache for frequently accessed tasks.
    """

    def __init__(self, repository: Optional[TaskRepository] = None) -> None:
        """
        Initialize the task service.

        Args:
            repository: Optional repository instance. Creates new one if not provided.
        """
        super().__init__()
        self.repository = repository or TaskRepository()
        self._cache: dict[UUID, Task] = {}
        self._cache_ttl = 300  # 5 minutes

    @log_call
    @measure_time
    def create_task(
        self,
        title: str,
        project_id: UUID,
        description: str = "",
        assignee_id: Optional[UUID] = None,
        priority: TaskPriority = TaskPriority.MEDIUM,
        due_date: Optional[datetime] = None,
        tags: Optional[list[str]] = None,
    ) -> Task:
        """
        Create a new task.

        Args:
            title: Task title.
            project_id: ID of the project the task belongs to.
            description: Detailed description.
            assignee_id: Optional user to assign.
            priority: Task priority level.
            due_date: Optional deadline.
            tags: Optional list of tags.

        Returns:
            The created task.

        Raises:
            ValidationError: If title is invalid.
        """
        is_valid, error_msg = validate_task_title(title)
        if not is_valid:
            raise ValidationError("title", error_msg)

        task = Task(
            title=title,
            project_id=project_id,
            description=description,
            assignee_id=assignee_id,
            priority=priority,
            due_date=due_date,
            tags=tags or [],
        )

        created = self.repository.create(task)
        self._log_info("Created task: %s", created.id)
        return created

    @log_call
    def get_task(self, task_id: UUID) -> Task:
        """
        Get a task by ID.

        Args:
            task_id: The task's unique identifier.

        Returns:
            The task.

        Raises:
            NotFoundError: If task doesn't exist.
        """
        task = self.repository.get(task_id)
        if not task:
            raise NotFoundError("Task", str(task_id))
        return task

    @log_call
    def update_task(
        self,
        task_id: UUID,
        title: Optional[str] = None,
        description: Optional[str] = None,
        status: Optional[TaskStatus] = None,
        priority: Optional[TaskPriority] = None,
        assignee_id: Optional[UUID] = None,
        due_date: Optional[datetime] = None,
    ) -> Task:
        """
        Update a task's properties.

        Args:
            task_id: The task to update.
            title: New title (if provided).
            description: New description (if provided).
            status: New status (if provided).
            priority: New priority (if provided).
            assignee_id: New assignee (if provided).
            due_date: New due date (if provided).

        Returns:
            The updated task.

        Raises:
            NotFoundError: If task doesn't exist.
            ValidationError: If new values are invalid.
        """
        task = self.get_task(task_id)

        if title is not None:
            is_valid, error_msg = validate_task_title(title)
            if not is_valid:
                raise ValidationError("title", error_msg)
            task.title = title

        if description is not None:
            task.description = description

        if status is not None:
            task.status = status

        if priority is not None:
            task.priority = priority

        if assignee_id is not None:
            task.assignee_id = assignee_id

        if due_date is not None:
            task.due_date = due_date

        task.updated_at = datetime.utcnow()
        self.repository.update(task)
        self._invalidate_cache(task_id)
        return task

    @log_call
    def delete_task(self, task_id: UUID) -> bool:
        """
        Delete a task.

        Args:
            task_id: The task to delete.

        Returns:
            True if deleted.

        Raises:
            NotFoundError: If task doesn't exist.
        """
        if not self.repository.exists(task_id):
            raise NotFoundError("Task", str(task_id))

        self.repository.delete(task_id)
        self._invalidate_cache(task_id)
        self._log_info("Deleted task: %s", task_id)
        return True

    async def create_task_async(
        self,
        title: str,
        project_id: UUID,
        **kwargs,
    ) -> Task:
        """
        Create a task asynchronously.

        Useful when creating tasks as part of a larger async workflow.

        Args:
            title: Task title.
            project_id: Project ID.
            **kwargs: Additional task properties.

        Returns:
            The created task.
        """
        # Simulate async operation
        await asyncio.sleep(0.01)
        return self.create_task(title, project_id, **kwargs)

    async def bulk_create_async(
        self,
        tasks_data: list[dict],
    ) -> list[Task]:
        """
        Create multiple tasks asynchronously.

        Args:
            tasks_data: List of task data dictionaries.

        Returns:
            List of created tasks.
        """
        tasks = []
        for data in tasks_data:
            task = await self.create_task_async(**data)
            tasks.append(task)
        return tasks

    async def get_tasks_with_refresh(
        self,
        project_id: UUID,
        force_refresh: bool = False,
    ) -> list[Task]:
        """
        Get tasks for a project with optional cache refresh.

        Args:
            project_id: Project to get tasks for.
            force_refresh: Whether to bypass cache.

        Returns:
            List of tasks.
        """
        if force_refresh:
            self._cache.clear()
            await asyncio.sleep(0.01)  # Simulate refresh delay

        return self.repository.get_by_project(project_id)

    def assign_task(self, task_id: UUID, user_id: UUID) -> Task:
        """
        Assign a task to a user.

        Args:
            task_id: The task to assign.
            user_id: The user to assign to.

        Returns:
            The updated task.
        """
        task = self.get_task(task_id)
        task.assign_to(user_id)
        self.repository.update(task)
        return task

    def complete_task(self, task_id: UUID) -> Task:
        """
        Mark a task as completed.

        Args:
            task_id: The task to complete.

        Returns:
            The updated task.
        """
        task = self.get_task(task_id)
        task.mark_complete()
        self.repository.update(task)
        return task

    def get_overdue_tasks(self) -> list[Task]:
        """Get all overdue tasks."""
        return self.repository.get_overdue()

    def get_tasks_by_status(self, status: TaskStatus) -> list[Task]:
        """Get tasks filtered by status."""
        return self.repository.get_by_status(status)

    def search_tasks(
        self,
        query: str,
        project_id: Optional[UUID] = None,
        status: Optional[TaskStatus] = None,
    ) -> list[Task]:
        """
        Search tasks with filters.

        Args:
            query: Search text.
            project_id: Optional project filter.
            status: Optional status filter.

        Returns:
            Matching tasks.
        """
        return self.repository.search(query, project_id, status)

    def _invalidate_cache(self, task_id: UUID) -> None:
        """Remove a task from cache."""
        self._cache.pop(task_id, None)

    @property
    def task_count(self) -> int:
        """Get total number of tasks."""
        return self.repository.count()

    @staticmethod
    def calculate_priority_score(task: Task) -> int:
        """
        Calculate a priority score for sorting.

        Args:
            task: The task to score.

        Returns:
            Priority score (higher = more important).
        """
        score = task.priority.value * 10

        if task.is_overdue:
            score += 50

        if task.status == TaskStatus.BLOCKED:
            score -= 20

        return score
